### PR TITLE
Import Results

### LINKED ISSUE #1502: Import Jupyter Notebook and create result cells
Creating an issue to track progress/thoughts on a possible "import notebook with results" command (it seems helpful for this to be a different issue+PR than "import notebook without results". (or a global setting for whether notebooks should be imported with results).

Result bubbles are usually created from the response of kernel execution:
https://github.com/nteract/hydrogen/blob/ea0004326a666c679119fbec582530123c4e7cca/lib/main.js#L393-L396

But all the responses from kernel execution _should_ also be stored in the Jupyter Notebook file. So it doesn't seem too crazy to think that it should be possible to import the results from the file and display them as it would from "run all cells".

### LINKED ISSUE #1552: Notebook Markdown import/export bug with consecutive newlines
Running the export on
```py
# %% markdown
# Hello world
#
# Test
# %%
print('hi')
```
gives cells of 
```json
{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "Hello world\n",
        "\nTest"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "source": [
        "print('hi')"
      ],
      "outputs": [],
      "execution_count": null,
      "metadata": {
        "collapsed": false,
        "outputHidden": false,
        "inputHidden": false
      }
    }
  ],
  "metadata": {},
  "nbformat": 4,
  "nbformat_minor": 4
}
```
The issue is that there's a leading `\n`. So when it gets imported back into Atom, the newline is for some reason stripped and no comment symbol is included:
```py
# %% markdown
# Hello world
# 
Test
# %%
print('hi')
```

I think it's standard in Jupyter notebook files to not have leading newlines; I think they're usually appended to the end. But not sure if that's a requirement or just common.

Either way current Hydrogen behavior is bad on import or export or both.

### PR DESCRIPTION
**Pull request by wadethestealth**
# 🌟 Importing Results
- Now on import of a jupyter notebook, results are rendered inline 😄 
- Fixes export/ import bug #1552

## To-Do:
- [x] add JSdoc comments
   - [x] import-notebook.js
   - [x] result.js
- [x] add Docs
- [ ] add Specs
---

### Notes
~To be merged **AFTER** #1714.~

Closes #1502 
Closes #1552

### CODE REVIEW

**Comment by wadethestealth** [comment]
Ping @aviatesk @BenRussert

File: lib/config.js [hunk 1]
```diff
@@ -47,6 +47,14 @@ const Config = {
       default: true,
       order: 3
     },
+    importNotebookResults: {
```
Comment (aviatesk) [inline lib/config.js]: Does this really need to be a setting ? If this works without bugs or some drawbacks, I don't this this can be default behaviour.

File: lib/import-notebook.js [hunk 2]
```diff
@@ -77,18 +113,34 @@ export async function _loadNotebook(filename: string) {
...
     return;
   }
-  const sources = _.map(nb.cells, cell => {
-    return getCellSource(cell, commentStartString + " ");
-  });
+  const nbCells = nb.cellMap.valueSeq().toArray();
+  let sources = [];
+  let resultRows = [];
+  let previousBreakpoint = -1;
+  for (let cell of nbCells) {
```
Comment (aviatesk) [inline lib/import-notebook.js]: I think we usually use `forEach` for for loops without break/return statements inside, thus I believe it's more preferable here too.

If you stick to use native `for`, I prefer `const` over `let` for `cell` 's identifier

File: lib/import-notebook.js [hunk 3]
```diff
@@ -47,25 +69,39 @@ export function importNotebook(event?: atom$CustomEvent) {
...
+    nb = fromJS(JSON.parse(data));
   } catch (err) {
     if (err.name === "SyntaxError") {
       atom.notifications.addError("Error not a valid notebook", {
         detail: err.stack
       });
     } else {
       atom.notifications.addError("Error reading file", {
-        detail: err.message
+        detail: err
```
Comment (aviatesk) [inline lib/import-notebook.js]: What's the content of `err` ?
If it's markdown string, we may want to use `description` instead of `detail`.

File: lib/import-notebook.js [hunk 4]
```diff
@@ -144,18 +208,88 @@ function getGrammarForKernelspecName(name: string): ?atom$Grammar {
...
+  resultRows: Array<number>
+) {
+  if (nbCells.length != resultRows.length) return;
+  let markers = store.markersMapping.get(editor.id);
+  markers = markers ? markers : store.newMarkerStore(editor.id);
+  let cellNumber = 0;
+  for (let cell of nbCells) {
+    const row = resultRows[cellNumber];
+
+    switch (cell.cell_type) {
```
Comment (aviatesk) [inline lib/import-notebook.js]: We don't need branch for `cell_type == "raw"` ?

File: lib/config.js (hunk 1 above)
Comment (wadethestealth) [reply lib/config.js]: I am sorry but I dont understand what that means. the reason for the setting is because it is rendered inline not the output pane and this could be undesired for many reasons so I made it optional. Also there are no draw backs and there should be very little impact since the only real change is adding a result bubble to the normal import procedure. I have no idea what bug could possibly prevent this from being a default implementation.

File: lib/import-notebook.js (hunk 3 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: I couldn't know since its a generic catch all.

File: lib/import-notebook.js (hunk 4 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: I will look into this.

**Comment by wadethestealth** [comment]
Also I will try to come up with some specs.

File: lib/config.js (hunk 1 above)
Comment (aviatesk) [reply lib/config.js]: Ah I'm sorry, horrible typos.
I tried to say "I think this can be default behavior" in the last sentence.

If you don't come up with any drawback, then let's make this the default and remove the config.


I feel like it would be better to keep configs clean as possible, since we have rather many.

Comment (aviatesk) [reply lib/config.js]: If an user don't like this, s/he can just
clear results.

And more, importing a notebook won't happen so much.

File: lib/import-notebook.js (hunk 4 above)
Comment (aviatesk) [reply lib/import-notebook.js]: Well, we still don't support exporting/rendering raw cells, thus it would be okay to just import raw cells as is for now :)

File: lib/import-notebook.js (hunk 3 above)
Comment (aviatesk) [reply lib/import-notebook.js]: So, you found some cases that `err.message` is undefined or something like that ?

Comment (wadethestealth) [reply lib/import-notebook.js]: I am pretty sure it was there before me I just removed the message portion because err should have a built-in to string that represents what the user should see,

File: lib/config.js (hunk 1 above)
Comment (wadethestealth) [reply lib/config.js]: Maybe you would understand better after local test because I could very easily see this as an annoyance. I know our config is a little messy, but I would rather clean that up rather than leave out a customization option.

Comment (aviatesk) [reply lib/config.js]: Okay sounds fair.
How about making a separate command for this feature ? i.e.: `Hydrogen: Import-Notebook-With-Results` will behaves like `importNotebookResults: true` while `Hydrogen: Import-Notebook` keeps behaving the same way as before.

I personally don't like changing configs so often and think using different command is easier.

File: lib/import-notebook.js [hunk 5]
```diff
@@ -77,18 +113,34 @@ export async function _loadNotebook(filename: string) {
   const editor = await atom.workspace.open();
   const grammar = getGrammarForNotebook(nb);
   if (!grammar) return;
-  editor.setGrammar(grammar);
+  atom.grammars.assignLanguageMode(editor, grammar.scopeName);
```
Comment (aviatesk) [inline lib/import-notebook.js]: What's the reason for this change ?
It seems to fail to recognise Julia notebook as julia file in my local test.

**Comment by aviatesk** [comment]
If you find writing specs difficult, it's okay to leave it to future :)

**Comment by aviatesk** [comment]
I found sometimes cells are ordered incorrectly on this PR, while this doesn't not happen on master.

Maybe the logic around this is kinda broken ?:

https://github.com/nteract/hydrogen/blob/dbef962ac7403471df96e836f5bf0d85f4feeee1/lib/import-notebook.js#L122-L132

This _seems_ to happen when the cells are executed randomly (i.e. not run from top to bottom) or including plots.

***

As a further question, what might happen when the original notebook includes cells that can't be rendered in Hydrogen, like ipywidgets or something like that ?
This might relate to the raw cells stuff, but I suspect some handling of those cases might be preferable here.

File: lib/import-notebook.js (hunk 5 above)
Comment (aviatesk) [reply lib/import-notebook.js]: Hmm, but I found the issue only happens on this PR, and not on master thus I suspect the change is causing something bad.

File: lib/config.js (hunk 1 above)
Comment (aviatesk) [reply lib/config.js]: We can register those commands on `atom-workspace` and then we can call them from tree-view context menus of course ?

Comment (aviatesk) [reply lib/config.js]: Just add the new command to import with results here ...
https://github.com/nteract/hydrogen/blob/bd951e8a471dea932f5f5c7f0c27e8936f71a106/menus/hydrogen.cson#L3

File: lib/import-notebook.js (hunk 5 above)
Comment (aviatesk) [reply lib/import-notebook.js]: Maybe for `getGrammarForNotebook` work, we still need the `setGrammar`.
You didn't face the issue maybe because Python grammar is handled with hard code.

***

Rather off topic, but I would really really
appreciate if you also test with various kernels other than IPython in general.

File: lib/config.js (hunk 1 above)
Comment (wadethestealth) [reply lib/config.js]: That won't allow you to set import with results to default though. But I agree it should still be added there.

File: lib/import-notebook.js (hunk 5 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: You are probably right and I believe that getGrammarForNotebook may need to be updated too.

**Comment by wadethestealth** [comment]
Update: I have fixed cell ordering, and it had nothing to do with execution, but rather I was using commutable wrong as cells are stored in a HashMap so there ordering is not guaranteed, which is why they also supply a List with the key order you should retrieve in.

File: lib/import-notebook.js (hunk 5 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: Fixed

File: lib/config.js (hunk 1 above)
Comment (wadethestealth) [reply lib/config.js]: I firmly believe this should be a config because we cannot register two file openers when left clicking a file to open it, and therefore I am leaving this as implemented

Comment (aviatesk) [reply lib/config.js]: How about creating two commands for importing notebooks, and register them in package menus ?
Then we can right-click on tree-view and select either way: https://github.com/nteract/hydrogen/blob/master/menus/hydrogen.cson#L1

The default behaviour (which is invoked by left-click) can be configured via config tho.

Comment (wadethestealth) [reply lib/config.js]: @aviatesk 
Roleplay:
So lets say I set default behavior with the config to be importing with results. Then the left click tree view uses that no problem and follows directions, but then what if I want to import via command palatte.
So I go and I run `hydrogen:import-notebook` and something weird happened, I see no results and I am confused, because I have my config to import results. I then realize that I have to call a completely separate command `hydrogen:import-notebook-with-results` and I think that is weird.
End Roleplay: I think to simplify for the user, we should be consistent with which version we are using. If someone wishes to load them without results, but does not want to switch it in the config. They will just have to clear results.

Comment (wadethestealth) [reply lib/config.js]: Also I don't think that this is blocking, so even if we disagree now, after merging your could always open a new PR and show your use case. @aviatesk

Comment (aviatesk) [reply lib/config.js]: Fair points. Thanks for the detailed reply :)

> Also I don't think that this is blocking, so even if we disagree now, after merging your could always open a new PR and show your use case. @aviatesk

yep, definitely. But actually I don't have such a strong opinion on this, I just wanted to make a suggestion ?

So let's resolve this as is.

File: lib/import-notebook.js [hunk 6]
```diff
@@ -47,89 +69,175 @@ export function importNotebook(event?: atom$CustomEvent) {
...
+    kernel_info, // Official nbformat v4
+    language, // Sometimes used in nbformat v3
+    language_info // Offical nbformat v4
+  } = nb.metadata.toJS();
+  const kernel = kernelspec ? kernelspec : kernel_info;
+  const lang = language_info
+    ? language_info
+    : language
+    ? { name: language }
+    : null;
```
Comment (aviatesk) [inline lib/import-notebook.js]:
```js
const lang = language_info ?
  language_info :
  language ?
    { name: language } :
    null
```

would look better and be consistent with the other parts of our code

File: lib/import-notebook.js [hunk 7]
```diff
@@ -47,89 +69,175 @@ export function importNotebook(event?: atom$CustomEvent) {
...
+  for (let g of grammars) {
+    if (
+      g &&
+      ((g.name && g.name.toLowerCase() == name.toLowerCase()) ||
+        g.scopeName == scopeName)
+    )
+      return g;
+  }
+  return null;
+}
```
Comment (aviatesk) [inline lib/import-notebook.js]: Why not just use [`atom.grammars.grammarForScopeName`](https://flight-manual.atom.io/api/v1.9.4/GrammarRegistry/#instance-grammarForScopeName) ?

**Comment by aviatesk** [comment]
@wadethestealth 
Do you think this PR is good to be merged as is ? If so, I'm going to merge this and let people try this out. Or you want me to do some local tests ?

**Comment by wadethestealth** [comment]
@aviatesk I say go ahead and merge but I need you to test raw cells, I don't know how to use them and don't have a notebook with them. Just make sure nothing crazy happens

**Comment by wadethestealth** [comment]
I will update the specs for this in a newer PR

File: lib/import-notebook.js (hunk 6 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: Agreed that was probably an auto lint as that looks nothing like i would do.

File: lib/import-notebook.js (hunk 7 above)
Comment (wadethestealth) [reply lib/import-notebook.js]: this way I only have to iterate once for both name and scope checking

**Comment by wadethestealth** [comment]
Going to merge this, and if any bugs come in I will do quick fix + specs and release a patch
